protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java

```diff
@@ -75,7 +79,41 @@ public synchronized void clear() {
     }
 
     public synchronized void addUid(MessageUid uid) {
-        this.addMapping(nextMsn(), uid);
+        if (isLastUid(uid)) {
+            addMapping(nextMsn(), uid);
+        } else {
+            addUidInMiddle(uid);
+        }
+    }
+
+    private boolean isLastUid(MessageUid uid) {
+        return msnToUid.isEmpty() ||
+            uid.asLong() > msnToUid.get(getLastMsn()).asLong();
```

Comment:
I won't agree on that. compareTo is a very common interface in Java. Using primitive types here is just pure duplication of logic that belongs to another type.